cabal-install/Distribution/Client/InstallPlan.hs

```diff
@@ -562,12 +565,12 @@ failed plan (Processing processingSet completedSet failedSet) pkgid =
     assert (all (`Set.notMember` processingSet) (tail newlyFailedIds)) $
     assert (all (`Set.notMember` completedSet)  (tail newlyFailedIds)) $
     assert (all (`Set.notMember` failedSet)     (tail newlyFailedIds)) $
-    assert (processingInvariant plan processing') $
+    assertMaybeMsg (processingInvariant plan processing') $
 
     ( map asConfiguredPackage (tail newlyFailed)
     , processing' )
   where
-    processingSet' = Set.delete pkgid processingSet
+    processingSet' = processingSet `Set.difference` Set.fromList newlyFailedIds
```

Comment:
Ah, I suspect that is a fix for a bug that was triggering the assertion failure...

Now that I see it, it's clear, we don't just remove the single failed package from the processing set but all the failed ones (the failure + the things that depend on the one that failed).

But yes, I expect @ezyang meant to put that fix in a separate patch.